src/sensor_hx71x.c

```diff
@@ -0,0 +1,267 @@
+    sched_add_timer(&hx71x->timer);
+    irq_enable();
+}
+
+int8_t
+hx71x_is_data_ready(struct hx71x_adc *hx71x) {
+    // if the dout pin is high the sample is not ready
+    return !gpio_in_read(hx71x->dout);
+}
+
+// Add a measurement to the buffer
+static void
+add_sample(struct hx71x_adc *hx71x, uint_fast32_t counts)
+{
+    hx71x->sb.data[hx71x->sb.data_count] = counts;
+    hx71x->sb.data[hx71x->sb.data_count + 1] = counts >> 8;
+    hx71x->sb.data[hx71x->sb.data_count + 2] = counts >> 16;
+    hx71x->sb.data[hx71x->sb.data_count + 3] = counts >> 24;
+    hx71x->sb.data_count += BYTES_PER_SAMPLE;
+}
+
+static void
+flush_samples(struct hx71x_adc *hx71x, uint8_t oid)
+{
+    if (hx71x->sb.data_count + BYTES_PER_SAMPLE > ARRAY_SIZE(hx71x->sb.data))
+        sensor_bulk_report(&hx71x->sb, oid);
+}
+
+// Pulse all clock pins to move to the next bit
+inline static void
+hx71x_pulse_clocks(struct hx71x_adc *hx71x) {
+    irq_disable();
+    gpio_out_write(hx71x->sclk, 1);
+    hx71x_delay_no_irq(hx71x_get_time(), MIN_PULSE_TIME);
+    gpio_out_write(hx71x->sclk, 0);
+    hx71x_delay_no_irq(hx71x_get_time(), MIN_PULSE_TIME);
+    irq_enable();
+}
+
+// hx71x ADC query
+void
+hx71x_read_adc(struct hx71x_adc *hx71x, uint8_t oid)
+{
+    if (!hx71x_is_data_ready(hx71x)) {
+        hx71x_reschedule_timer(hx71x);
+        return;
+    }
+
+    // data is ready
+    int32_t counts = 0;
+    hx71x_time_t start_time = timer_read_time();
+    for (uint_fast8_t sample_idx = 0; sample_idx < 24; sample_idx++) {
+        hx71x_pulse_clocks(hx71x);
+        // read 2's compliment int bits
+        counts = (counts << 1) | gpio_in_read(hx71x->dout);
+    }
+
+    // bit bang 1 to 4 more bits to configure gain & channel for the next sample
+    for (uint8_t gain_idx = 0; gain_idx < hx71x->gain_channel; gain_idx++) {
+        hx71x_pulse_clocks(hx71x);
+    }
+
+    if (hx71x_is_data_ready(hx71x)) {
+        shutdown("HX71x Sensor reporting data ready after read");
+    }
+
+    hx71x_time_t time_diff = timer_read_time() - start_time;
+    if (time_diff >= MAX_READ_TIME) {
+        shutdown("HX71x Read took too long");
+    }
+
+    // extend 2's complement 24 bits to 32bits
+    if (counts >= 0x800000) {
+        counts |= 0xFF000000;
+    }
+
+    if (counts < -0x7FFFFF || counts > 0x7FFFFF) {
+        shutdown("HX71x Counts value out of range");
+    }
+    add_sample(hx71x, counts);
+
+    flush_samples(hx71x, oid);
+    hx71x_reschedule_timer(hx71x);
+}
+
+// Create a hx71x sensor
+void
+command_config_hx71x(uint32_t *args)
+{
+    struct hx71x_adc *hx71x = oid_alloc(args[0]
+                , command_config_hx71x, sizeof(*hx71x));
+    hx71x->timer.func = hx71x_event;
+    hx71x->pending_flag = 0;
+    uint8_t gain_channel = args[1];
+    if (gain_channel < 1 || gain_channel > 4) {
+        shutdown("HX71x gain/channel out of range 1-4");
+    }
+    hx71x->gain_channel = gain_channel;
+    hx71x->dout = gpio_in_setup(args[2], -1);
```

Comment:
A gpio pull down is not supported on all chips.  So, this look strange.  Could a pull up be used instead?

Reply:
Looks like a pull-up works (on my hardware at least)